nexus_constructor/geometry/off_geometry.py

```diff
@@ -0,0 +1,233 @@
+)
+import numpy as np
+
+
+class OFFGeometry(ABC):
+    geometry_str = "OFF"
+
+    def __init__(self, units: str = "", filename: str = ""):
+        """
+        :param units: Store units from interface to repopulate field if component is edited
+        :param filename: Store filename from interface to repopulate field if component is edited
+        """
+        self.units = units
+        self.filename = filename
+
+    @property
+    @abstractmethod
+    def winding_order(self):
+        """
+        Flattened 1D list of indices in vertices for each face
+        winding_order_indices gives the start index for each face in this list
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def winding_order_indices(self):
+        """
+        The start index for each face in winding_order
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def off_geometry(self):
+        pass
+
+    @property
+    @abstractmethod
+    def vertices(self) -> List[QVector3D]:
+        pass
+
+    @vertices.setter
+    @abstractmethod
+    def vertices(self, new_vertices: List[QVector3D]):
+        pass
+
+    @property
+    @abstractmethod
+    def faces(self) -> List[List[int]]:
+        pass
+
+    @faces.setter
+    @abstractmethod
+    def faces(self, new_faces: List[List[int]]):
+        pass
+
+
+class OFFGeometryNoNexus(OFFGeometry):
+    """
+    3D mesh description of the shape of an object, based on the OFF file format.
+    This class does not store its data in the NeXus file, used for placeholder shape
+    for objects which have no real shape data to be stored in the file.
+    """
+
+    def __init__(self, vertices: List[QVector3D] = None, faces: List[List[int]] = None):
+        """
+        :param vertices: list of Vector objects used as corners of polygons in the geometry
+        :param faces: list of integer lists. Each sublist is a winding path around the corners of a polygon.
+            Each sublist item is an index into the vertices list to identify a specific point in 3D space
+        """
+        super().__init__()
+        self._vertices = vertices
+        self._faces = faces
+
+    @property
+    def winding_order(self):
+        return [point for face in self.faces for point in face]
+
+    @property
+    def winding_order_indices(self):
+        face_sizes = [len(face) for face in self.faces]
+        return [sum(face_sizes[0:i]) for i in range(len(face_sizes))]
+
+    @property
+    def off_geometry(self) -> OFFGeometry:
+        return self
+
+    @property
+    def vertices(self) -> List[QVector3D]:
+        return self._vertices
+
+    @vertices.setter
+    def vertices(self, new_vertices: List[QVector3D]):
+        self._vertices = new_vertices
+
+    @property
+    def faces(self) -> List[List[int]]:
+        return self._faces
+
+    @faces.setter
+    def faces(self, new_faces: List[List[int]]):
+        self._faces = new_faces
+
+
+class OFFGeometryNexus(OFFGeometry):
```

Comment:
Can you add some unit tests to cover this class in general?

Reply:
It is tested in `test_off_geometry.py`. Maybe it isn't obvious because I populate it by using an `OFFGeometryNoNexus`?

Reply:
Added several more tests to `test_off_geometry`